src/Chainweb/BlockHeader.hs

```diff
@@ -468,150 +617,140 @@ instance HasChainGraph ParentHeader where
-            -- ^ authoritative
-
-        , _blockChainId :: {-# UNPACK #-} !ChainId
-
-        , _blockWeight :: {-# UNPACK #-} !BlockWeight
-            -- ^ the accumulated weight of the chain. It is redundant information
-            -- that is subject to the inductive property that the block weight
-            -- of a block is the block weight of the parent plus the difficulty
-            -- of the block.
-
-        , _blockHeight :: {-# UNPACK #-} !BlockHeight
-            -- ^ block height records the length of the chain. It is redundant
-            -- information and thus subject the inductive property that
-            -- the block height of a block is the block height of its parent
-            -- plus one.
-
-        , _blockChainwebVersion :: !ChainwebVersion
-            -- ^ the Chainweb version is a constant for the chain. A chain
-            -- is uniquely identified by its genesis block. Thus this is
-            -- redundant information and thus subject to the inductive property
-            -- that the Chainweb version of a block equals the Chainweb version
-            -- of its parent.
-
-        , _blockEpochStart :: {-# UNPACK #-} !EpochStartTime
-            -- ^ The start time of the current difficulty adjustment epoch.
-            -- Epochs divide the sequence of blocks in the chain into continuous
-            -- ranges of blocks. Each epoch is defined by the minimal block
-            -- height of the blocks in the epoch.
-
-        , _blockNonce :: {-# UNPACK #-} !Nonce
-            -- ^ authoritative
-
-        , _blockHash :: {-# UNPACK #-} !BlockHash
-            -- ^ the hash of the block. It includes all of the above block properties.
-        }
-        -> BlockHeader
-        deriving (Show, Generic)
-        deriving anyclass (NFData)
-
 isGenesisBlockHeader :: BlockHeader -> Bool
 isGenesisBlockHeader b =
-    _blockHeight b == genesisHeight (_blockChainwebVersion b) (_blockChainId b)
-{-# INLINE isGenesisBlockHeader #-}
+    _blockHeight b == genesisHeight (_chainwebVersion b) (_chainId b)
 
-instance Eq BlockHeader where
-     (==) = (==) `on` _blockHash
-     {-# INLINE (==) #-}
+-- | The genesis block hash includes the Chainweb version and the 'ChainId'
+-- within the Chainweb version.
+--
+-- It is the '_blockParent' of the genesis block
+--
+genesisParentBlockHash :: HasChainId p => ChainwebVersion -> p -> BlockHash
+genesisParentBlockHash v p = BlockHash $ MerkleLogHash
+    $ merkleRoot $ merkleTree @ChainwebMerkleHashAlgorithm
+        [ InputNode "CHAINWEB_GENESIS"
+        , encodeMerkleInputNode encodeChainwebVersionCode (_versionCode v)
+        , encodeMerkleInputNode encodeChainId (_chainId p)
+        ]
 
-instance Ord BlockHeader where
-     compare = compare `on` _blockHash
-     {-# INLINE compare #-}
+{-# NOINLINE genesisBlockHeaderCache #-}
+genesisBlockHeaderCache :: IORef (HashMap ChainwebVersionCode (HashMap ChainId BlockHeader))
```

Comment:
We may consider doing a more efficient memoization for production versions (in particular with large histories, where we may do long replays or catchups).